python/src/search_algorithm.py

```diff
@@ -36,6 +37,49 @@ def get_next_population(
         pass
 
 
+@dataclass()
+class CandidateCache(Generic[Candidate]):
+    _size: int
+    _cache: Dict[int, float] = field(default_factory=dict, init=False)
+    _hits: int = field(default=0, init=False)
+    _misses: int = field(default=0, init=False)
+
+    def __post_init__(self):
+        assert self._size > 0
+
+    def get(self, candidates: Sequence[Candidate]) -> Tuple[CostMap, Sequence[Candidate]]:
+        """Returns costs for cached candidates and misses"""
+        # Get cached costs by hash on candidate
+        cached_cost_map: CostMap = {
+            c.key(): self._cache[hash(c)]
+            for c in candidates
+            if hash(c) in self._cache
+        }
+        self._hits += len(cached_cost_map)
+
+        # Get the others by fetch function (probably calls a simulator and cost function)
+        uncached_candidates = tuple(filter(lambda c: c.key() not in cached_cost_map, candidates))
+        self._misses += len(uncached_candidates)
+        return cached_cost_map, uncached_candidates
+
+    def update(self, candidates: Sequence[Candidate], cost_map: CostMap):
+        # Make one big cache with existing and new entries, deduplicate using dict
+        combined_cache = {hash(c): cost_map[c.key()] for c in candidates}  # New entries
+        combined_cache.update(self._cache)  # Add existing
```

Comment:
For readability, can we change this to:
```python
new_entries = {hash(c): cost_map[c.key()] for c in candidates}
combined_cache = {**self._cache, **new_entries}
```
I just find it confusing that we create a dict called combined_cache that isn't combined at first. Obviously your comments call that out, but I think this might be a bit cleaner

Reply:
Sure